src/category_theory/abelian/non_preadditive.lean

```diff
@@ -0,0 +1,727 @@
+/-
+Copyright (c) 2020 Markus Himmel. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Markus Himmel
+-/
+import category_theory.limits.shapes.binary_products
+import category_theory.limits.shapes.kernels
+import category_theory.limits.shapes.pullbacks
+import category_theory.limits.shapes.regular_mono
+import category_theory.preadditive
+
+/-!
+# Every non_preadditive_abelian category is preadditive
+
+In mathlib, we define an abelian category as a preadditive category with a zero object,
+kernels and cokernels, products and coproducts and in which every monomorphism and epimorphis is
+normal.
+
+While virtually every interesting abelian category has a natural preadditive structure (which is why
+it is included in the definition), preadditivity is not actually needed: Every category that has
+all of the other properties appearing in the definition of an abelian category admits a preadditive
+structure. This is the construction we carry out in this file.
+
+The proof proceeds in roughly five steps:
+1. Prove some results (for example that all equalizers exist) that would be trivial if we already
+   had the preadditive structure but are a bit of work without it.
+2. Develop images and coimages to show that every monomorphism is the kernel of its cokernel.
+
+The results of the first two steps are also useful for the "normal" development of abelian
+categories, and will be used there.
+
+3. For every object `A`, define a "subtraction" morphism `σ : A ⨯ A ⟶ A` and use it to define
+   subtraction on morphisms as `f - g := prod.lift f g ≫ σ`.
+4. Prove a small number of identities about this subtraction from the definition of `σ`.
+5. From these identities, prove a large number of other identities that imply that defining
+   `f + g := f - (0 - g)` indeed gives an abelian group structure on morphisms such that composition
+   is bilinear.
+
+The construction is non-trivial and it is quite remarkable that this abelian group structure can
+be constructed purely from the existence of a few limits and colimits. What's even more impressive
+is that all additive structures on a category are in some sense isomorphic, so for abelian
+categories with a natural preadditive structure, this construction manages to "almost" reconstruct
+this natural structure. However, we have not formalized this isomorphism.
+
+## References
+
+* [F. Borceux, *Handbook of Categorical Algebra 2*][borceux-vol2]
+
+-/
+
+open category_theory
+open category_theory.limits
+
+namespace category_theory
+section
+universes v u
+
+variables (C : Type u) [category.{v} C]
+
+/-- In this file, we call a category non_preadditive_abelian if it has a zero object, kernels, cokernel, binary
```

Comment:
This shouldn't be a docstring, I think.
Is the plan to use this class outside of this file? Or do you want to limit it to this file?
In the latter case, maybe it is safer to do
```
structure non_preadditive_abelian := _

local attribute [class] non_preadditive_abelian
```

Reply:
The class can be used outside of the file, for example in `abelian.lean`.

Reply:
Ok, thanks for clarifying.
Should there be a
```
def abelian_of_nonadditive_abelian := _
```
that allows to construct an instance of `abelian C` from a nonpreadditive one?

Reply:
Yes. It is at the bottom of `abelian.lean`, for import ordrr reasons.